rye/src/cli/publish.rs

```diff
@@ -0,0 +1,210 @@
+use std::io::{Read, Write};
+use std::path::PathBuf;
+use std::process::{Command, Stdio};
+
+use age::{
+    secrecy::{ExposeSecret, Secret},
+    Decryptor, Encryptor,
+};
+use anyhow::{bail, Context, Error};
+use clap::Parser;
+use rpassword::read_password;
+use toml_edit::{Item, Table};
+
+use crate::bootstrap::ensure_self_venv;
+use crate::config::{get_credentials, write_credentials};
+use crate::pyproject::PyProject;
+use crate::utils::CommandOutput;
+
+/// Publish packages to a package repository.
+#[derive(Parser, Debug)]
+pub struct Args {
+    /// The distribution files to upload to the repository (defaults to <workspace-root>/dist/*).
+    dist: Option<Vec<PathBuf>>,
+    /// The repository to publish to (defaults to 'pypi').
+    #[arg(short, long, default_value = "pypi")]
+    repository: String,
+    /// The repository url to publish to (defaults to https://upload.pypi.org/legacy/).
+    #[arg(long, default_value = "https://upload.pypi.org/legacy/")]
+    repository_url: String,
```

Comment:
```suggestion
    #[arg(long, default_value = "https://upload.pypi.org/legacy/")]
    repository_url: Url,
```
This could probably work. I'll check this out tonight.

Reply:
This seems to work. 

I made a [comment about handling trailing slashes in the url](https://github.com/mitsuhiko/rye/pull/86#issuecomment-1540678253). If `twine` doesn't handle this upfront for the user maybe we don't either. Here's the output the user would see 

```
Enter a passphrase (optional): 
Uploading distributions to https://test.pypi.org/legacy
Uploading rye_publish-0.1.0-py3-none-any.whl
100% ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 kB • 00:00 • ?
ERROR    RedirectDetected: https://test.pypi.org/legacy attempted to redirect to                    
         https://test.pypi.org/legacy/.                                                             
         Your repository URL is missing a trailing slash. Please add it and try again.
```

Reply:
Added `Url` usage in ba19effff21e4e78fb9279e03a42e0768ab95016